weather_mv/loader_pipeline/bq.py

```diff
@@ -253,57 +329,69 @@ def extract_rows(self, uri: str, coordinates: t.List[t.Dict]) -> t.Iterator[t.Di
 
-    def to_rows(self, coordinates: t.Iterable[t.Dict], ds: xr.Dataset, uri: str) -> t.Iterator[t.Dict]:
+    def to_rows(self, coordinate: t.Dict, ds: xr.Dataset, uri: str) -> t.Iterator[t.Dict]:
         first_ts_raw = (
             ds.time[0].values if isinstance(ds.time.values, np.ndarray)
             else ds.time.values
         )
         first_time_step = to_json_serializable_type(first_ts_raw)
-        for it in coordinates:
-            # Use those index values to select a Dataset containing one row of data.
-            row_ds = ds.loc[it]
+        with open_local(self.geo_data_csv_path) as master_lat_lon:
+            selected_ds = ds.loc[coordinate]
 
-            # Create a Name-Value map for data columns. Result looks like:
-            # {'d': -2.0187, 'cc': 0.007812, 'z': 50049.8, 'rr': None}
-            row = {n: to_json_serializable_type(ensure_us_time_resolution(v.values))
-                   for n, v in row_ds.data_vars.items()}
+            # Ensure that the latitude and longitude dimensions are in sync with the geo data CSV.
+            if not BQ_EXCLUDE_COORDS - set(selected_ds.dims.keys()):
+                selected_ds = selected_ds.transpose('latitude', 'longitude')
 
-            # Serialize coordinates.
-            it = {k: to_json_serializable_type(v) for k, v in it.items()}
+            master_df = pd.read_csv(master_lat_lon)
+            if self.skip_creating_polygon:
+                master_df[GEO_POLYGON_COLUMN] = None
 
             # Add indexed coordinates.
-            row.update(it)
+            for k, v in coordinate.items():
+                master_df[k] = to_json_serializable_type(v)
+
             # Add un-indexed coordinates.
-            for c in row_ds.coords:
-                if c not in it and (not self.variables or c in self.variables):
-                    row[c] = to_json_serializable_type(ensure_us_time_resolution(row_ds[c].values))
-
-            # Add import metadata.
-            row[DATA_IMPORT_TIME_COLUMN] = self.import_time
-            row[DATA_URI_COLUMN] = uri
-            row[DATA_FIRST_STEP] = first_time_step
-
-            longitude = ((row['longitude'] + 180) % 360) - 180
-            row[GEO_POINT_COLUMN] = fetch_geo_point(row['latitude'], longitude)
-            row[GEO_POLYGON_COLUMN] = (
-                fetch_geo_polygon(row["latitude"], longitude, self.lat_grid_resolution, self.lon_grid_resolution)
-                if not self.skip_creating_polygon
-                else None
-            )
-            # 'row' ends up looking like:
-            # {'latitude': 88.0, 'longitude': 2.0, 'time': '2015-01-01 06:00:00', 'd': -2.0187, 'cc': 0.007812,
-            #  'z': 50049.8, 'data_import_time': '2020-12-05 00:12:02.424573 UTC', ...}
-            metric.Metrics.counter('Success', 'ExtractRows').inc()
-            yield row
+            # Filter out excluded coordinates from coords.
+            filtered_coords = [c for c in selected_ds.coords if c not in BQ_EXCLUDE_COORDS]
```

Comment:
Maybe you could do one less loop if this was a generator comprehension ().

Reply:
umm I did not get the idea here :). Can you please elaborate.

Reply:
Replace the `[…]` with `(…)` and it will become a generator expression (evaluated lazily).

Reply:
ah ! my bad ... got it. Changes done.